contracts/airdrops/shared/AirdropExistingToken.sol

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.15;
+
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "./AirdropTimeData.sol";
+
+abstract contract AirdropExistingToken is AirdropTimeData {
+    IERC721 public immutable rewardedNft;
+    uint256 public immutable tokensPerClaim;
+    uint256 public immutable totalAirdropAmount;
+
+    mapping(uint256 => bool) public hasClaimed;
+
+    uint256 public airdropFundBlockTimestamp;
+    bool public airdropFunded;
+
+    address internal airdropFundingHolder;
+
+    event Claimed(uint256 indexed tokenId, address indexed claimer);
+    event AirdropFunded(address contractAddress);
+
+    error AirdropStillInProgress();
+    error AlreadyRedeemed();
+    error AlreadyFunded();
+    error AirdropExpired();
+    error Unauthorized();
+
+    constructor(
+        address _rewardedNft,
+        uint256 _tokensPerClaim,
+        uint256 _totalAirdropAmount,
+        uint256 _airdropDuration
+    ) AirdropTimeData(_airdropDuration) {
+        rewardedNft = IERC721(_rewardedNft);
+        tokensPerClaim = _tokensPerClaim;
+        totalAirdropAmount = _totalAirdropAmount;
+    }
+
+    /// @notice Changes state and emits event once contracts have been funded with tokens.
+    /// A revert of the transfer of existing tokens to the contract will revert these changes.
+    /// @dev This method is called as super.fundAirdrop() in Existing1155NftDrop.sol,
+    /// and ExistingTokenDrop.sol in standard airdrops.
+    function fundAirdrop() public virtual {
+        if (airdropFunded) revert AlreadyFunded();
+
+        airdropFunded = true;
+        airdropFundBlockTimestamp = block.timestamp;
+        airdropFundingHolder = msg.sender;
+
+        emit AirdropFunded(address(this));
+    }
+
+    /// @notice Allows the airdrop funder to withdraw back their funds after the airdrop has finished
+    function withdrawAirdropFunds() public virtual {
+        if (airdropFundingHolder != msg.sender) revert Unauthorized();
+        if (block.timestamp < airdropFinishTime) revert AirdropStillInProgress();
+    }
+
+    /// @notice Allows the NFT holder to claim their ERC20 airdrop
+    /// @param tokenId is the rewarded NFT collections token ID
+    function claim(uint256 tokenId) public virtual {
+        validateClaim(tokenId);
+        if (block.timestamp > airdropFinishTime) revert AirdropExpired();
+
+        hasClaimed[tokenId] = true;
+        emit Claimed(tokenId, msg.sender);
+    }
+
+    /// @notice Claim multiple ERC20 airdrops at once
+    /// @param tokenIds are the rewarded NFT collections token ID's
```

Comment:
add @dev comment explaining that this function doesn't transfer token, and that it should be handled in child contract